File: agent_service/preference_learner.py
import os
import json
from typing import Optional


class PreferenceLearner:
    def __init__(self, save_path: str = "./preferences.json"):
        self.save_path = save_path
        self.data = self._load()

    def _load(self) -> dict:
        if os.path.exists(self.save_path):
            with open(self.save_path, "r", encoding="utf-8") as f:
                return json.load(f)
        return {
            "accepted": {},
            "rejected": {},
            "thresholds": {
                "link_confidence_min": 0.3,
                "split_min_chars": 800,
                "merge_max_chars": 100,
                "moc_cluster_min": 10,
                "orphan_link_max": 2,
            },
        }

    def _save(self):
        with open(self.save_path, "w", encoding="utf-8") as f:
            json.dump(self.data, f, ensure_ascii=False, indent=2)

    def record(self, action_type: str, suggestion: str, accepted: bool):
        key = "accepted" if accepted else "rejected"
        bucket = self.data[key]
        if action_type not in bucket:
            bucket[action_type] = []
        # 去重：同一条 suggestion 不重复记录
        if suggestion in bucket[action_type]:
            return
        bucket[action_type].append(suggestion)
        if len(bucket[action_type]) > 100:
            bucket[action_type] = bucket[action_type][-100:]
        self._save()

    def get_accept_rate(self, action_type: str) -> Optional[float]:
        accepted = len(self.data["accepted"].get(action_type, []))
        rejected = len(self.data["rejected"].get(action_type, []))
        total = accepted + rejected
        if total == 0:
            return None
        return accepted / total

    def get_thresholds(self) -> dict:
        return self.data["thresholds"]

    def update_threshold(self, key: str, value: float):
        self.data["thresholds"][key] = value
        self._save()


preference_learner = PreferenceLearner()

File: obsidian-plugin/src/preference_learner.ts
/**
 * 偏好学习器
 * 记录用户的采纳/拒绝行为，计算采纳率，管理阈值。
 * 使用 Obsidian Plugin 的 loadData/saveData 持久化。
 */

export interface PreferenceData {
  accepted: Record<string, string[]>;
  rejected: Record<string, string[]>;
  thresholds: {
    link_confidence_min: number;
    split_min_chars: number;
    merge_max_chars: number;
    moc_cluster_min: number;
    orphan_link_max: number;
  };
}

const DEFAULT_DATA: PreferenceData = {
  accepted: {},
  rejected: {},
  thresholds: {
    link_confidence_min: 0.3,
    split_min_chars: 800,
    merge_max_chars: 100,
    moc_cluster_min: 10,
    orphan_link_max: 2,
  },
};

export class PreferenceLearner {
  private data: PreferenceData;
  private saveFn: (data: PreferenceData) => void;

  constructor(
    saveFn: (data: PreferenceData) => void,
    initialData?: Partial<PreferenceData>
  ) {
    this.saveFn = saveFn;
    this.data = { ...DEFAULT_DATA, ...initialData };
    if (initialData?.thresholds) {
      this.data.thresholds = { ...DEFAULT_DATA.thresholds, ...initialData.thresholds };
    }
    if (initialData?.accepted) {
      this.data.accepted = initialData.accepted;
    }
    if (initialData?.rejected) {
      this.data.rejected = initialData.rejected;
    }
  }

  /**
   * 记录一次反馈
   */
  record(actionType: string, suggestion: string, accepted: boolean): void {
    const key = accepted ? "accepted" : "rejected";
    const bucket = this.data[key];

    if (!bucket[actionType]) {
      bucket[actionType] = [];
    }

    // 去重
    if (bucket[actionType].includes(suggestion)) return;

    bucket[actionType].push(suggestion);

    // 保留最近 100 条
    if (bucket[actionType].length > 100) {
      bucket[actionType] = bucket[actionType].slice(-100);
    }

    this.saveFn(this.data);
  }

  /**
   * 获取某类建议的历史采纳率
   */
  getAcceptRate(actionType: string): number | null {
    const accepted = this.data.accepted[actionType]?.length || 0;
    const rejected = this.data.rejected[actionType]?.length || 0;
    const total = accepted + rejected;
    if (total === 0) return null;
    return accepted / total;
  }

  getThresholds(): PreferenceData["thresholds"] {
    return this.data.thresholds;
  }

  updateThreshold(key: keyof PreferenceData["thresholds"], value: number): void {
    this.data.thresholds[key] = value;
    this.saveFn(this.data);
  }

  /**
   * 获取数据（用于注入到 Agent prompt）
   */
  getData(): PreferenceData {
    return this.data;
  }
}
